src/rgw/rgw_rest_swift.cc

```diff
@@ -41,6 +41,110 @@
 
 using namespace std;
 
+template <class H, bool O>
+class FormPostSignatureT: public rgw::auth::swift::FormatSignature<H,O>
+{
+  using UCHARPTR = const unsigned char*;
+  using base_t = rgw::auth::swift::SignatureHelperT<H>;
+  using format_signature_t = rgw::auth::swift::FormatSignature<H,O>;
+public:
+  const char* calc(const std::string& key,
+      const std::string_view& path_info,
+      const std::string_view& redirect,
+      const std::string_view& max_file_size,
+      const std::string_view& max_file_count,
+      const std::string_view& expires) {
+    H hmac((UCHARPTR) key.data(), key.size());
+
+    hmac.Update((UCHARPTR) path_info.data(), path_info.size());
+    hmac.Update((UCHARPTR) "\n", 1);
+
+    hmac.Update((UCHARPTR) redirect.data(), redirect.size());
+    hmac.Update((UCHARPTR) "\n", 1);
+
+    hmac.Update((UCHARPTR) max_file_size.data(), max_file_size.size());
+    hmac.Update((UCHARPTR) "\n", 1);
+
+    hmac.Update((UCHARPTR) max_file_count.data(), max_file_count.size());
+    hmac.Update((UCHARPTR) "\n", 1);
+
+    hmac.Update((UCHARPTR) expires.data(), expires.size());
+
+    hmac.Final(base_t::dest);
+
+    return format_signature_t::result();
+  }
+};
+class RGWFormPost::SignatureHelper {
+public:
+  virtual ~SignatureHelper() {};
+  virtual const char* calc(const std::string& key,
+    const std::string_view& path_info,
+    const std::string_view& redirect,
+    const std::string_view& max_file_size,
+    const std::string_view& max_file_count,
+    const std::string_view& expires) {
+    return nullptr;
+  };
+  virtual const char* get_signature() const {
+    return nullptr;
+  };
+  virtual bool is_equal_to(const std::string& rhs) {
+    return false;
+  };
+  static SignatureHelper* get_sig_helper(std::string_view x);
+};
+template<typename H, bool O>
+class RGWFormPost::SignatureHelper_x : public RGWFormPost::SignatureHelper
+{
+  friend RGWFormPost;
+private:
+  FormPostSignatureT<H,O> d;
+protected:
+  ~SignatureHelper_x() { };
+  SignatureHelper_x() {};
+public:
+  virtual const char* calc(const std::string& key,
+    const std::string_view& path_info,
+    const std::string_view& redirect,
+    const std::string_view& max_file_size,
+    const std::string_view& max_file_count,
+    const std::string_view& expires) {
+    return d.calc(key,path_info,redirect,
+      max_file_size,max_file_count,expires) ;
+  };
+  virtual const char* get_signature() const {
+    return d.get_signature();
+  };
+  virtual bool is_equal_to(const std::string& rhs) {
+    return d.is_equal_to(rhs);
+  };
+};
+
+RGWFormPost::SignatureHelper* RGWFormPost::SignatureHelper::get_sig_helper(std::string_view x) {
+  size_t pos = x.find(':');
+  if (pos == x.npos || pos <= 0) {
+    switch(x.length()) {
+    case CEPH_CRYPTO_HMACSHA1_DIGESTSIZE*2:
```

Comment:
why hasn't the template split removed the need for the switch statement?  i.e., you've templated out the HMAC and (um, O) parameters in descendant template types.  Could the ancestor types become abstract, maybe?

Reply:
The switch statement here is part of the logic that decides at run-time which of the 6 different forms of template logic it should realize.  All the template logic happens strictly at compile-time, it can't replace the ultimate need to do grody old-fashioned run-time string parsing, which is what's happening here.
  There are however 2 things I didn't do in this direction: firrstly, there are 2 copies of get_sig_helper, one for tempurl and one for formpost.  It could  be trivially one templated function, or less trivially, integrated into some other base class.  There's no runtime advantage to this, and it makes include files a bit more messy.
  The other thing is the template classes RGWFormPost::SignatureHelper_x  and TempURLSignature::SignatureHelper_x, these both mainly contain thunks for calc() and is_equal_to().  Slightly more clever class logic might eliminate the need for those, but I kept ending up with worse kludges chasing that down.

Reply:
I like the use of templates, sorry I seemed to be over reading here;  I would rather we'd landed this in 22, rather than risking it getting stuck